File: bindings/kotlin/src/main/kotlin/strling/core/IR.kt
package strling.core

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonClassDiscriminator
import kotlinx.serialization.json.JsonElement

/**
 * STRling Intermediate Representation (IR) Node Definitions
 * 
 * This module defines the complete set of IR node classes that represent
 * language-agnostic regex constructs.
 */

// ---- Base IR operation ----

@Serializable
@JsonClassDiscriminator("ir")
sealed interface IROp

@Serializable
@JsonClassDiscriminator("ir")
sealed interface IRClassItem

// ---- IR concrete operations ----

@Serializable
@SerialName("Alt")
data class IRAlt(val branches: List<IROp>) : IROp

@Serializable
@SerialName("Seq")
data class IRSeq(val parts: List<IROp>) : IROp

@Serializable
@SerialName("Lit")
data class IRLit(val value: String) : IROp

@Serializable
@SerialName("Dot")
class IRDot : IROp {
    override fun equals(other: Any?): Boolean = other is IRDot
    override fun hashCode(): Int = "IRDot".hashCode()
}

@Serializable
@SerialName("Anchor")
data class IRAnchor(val at: String) : IROp

// ---- IR CharClass ----

@Serializable
@SerialName("CharClass")
data class IRCharClass(val negated: Boolean, val items: List<IRClassItem>) : IROp

@Serializable
@SerialName("Range")
data class IRClassRange(val from: String, val to: String) : IRClassItem

@Serializable
@SerialName("Char")
data class IRClassLiteral(val char: String) : IRClassItem

@Serializable
@SerialName("Esc")
data class IRClassEscape(val type: String, val property: String? = null) : IRClassItem

@Serializable
@SerialName("Quant")
data class IRQuant(
    val child: IROp,
    val min: Int,
    val max: JsonElement, // Int or "Inf"
    val mode: String
) : IROp

@Serializable
@SerialName("Group")
data class IRGroup(
    val capturing: Boolean,
    val body: IROp,
    val name: String? = null,
    val atomic: Boolean? = null
) : IROp

@Serializable
@SerialName("Backref")
data class IRBackref(
    val byIndex: Int? = null,
    val byName: String? = null
) : IROp

@Serializable
@SerialName("Look")
data class IRLook(
    val dir: String,
    val neg: Boolean,
    val body: IROp
) : IROp

File: bindings/php/src/Core/IR.php
<?php

namespace STRling\Core;

use JsonSerializable;

interface IRNode extends JsonSerializable {}

readonly class Lit implements IRNode
{
    public function __construct(
        public string $value
    ) {}

    public function jsonSerialize(): mixed
    {
        return [
            'ir' => 'Lit',
            'value' => $this->value,
        ];
    }
}

readonly class Seq implements IRNode
{
    /** @param IRNode[] $parts */
    public function __construct(
        public array $parts
    ) {}

    public function jsonSerialize(): mixed
    {
        return [
            'ir' => 'Seq',
            'parts' => $this->parts,
        ];
    }
}

readonly class Quant implements IRNode
{
    public function __construct(
        public IRNode $child,
        public int $min,
        public string|int $max, // "Inf" or int
        public string $mode // "Greedy", "Lazy", "Possessive"
    ) {}

    public function jsonSerialize(): mixed
    {
        return [
            'ir' => 'Quant',
            'child' => $this->child,
            'min' => $this->min,
            'max' => $this->max,
            'mode' => $this->mode,
        ];
    }
}

readonly class CharClass implements IRNode
{
    /** @param IRNode[] $items */
    public function __construct(
        public bool $negated,
        public array $items
    ) {}

    public function jsonSerialize(): mixed
    {
        return [
            'ir' => 'CharClass',
            'negated' => $this->negated,
            'items' => $this->items,
        ];
    }
}

readonly class Char implements IRNode
{
    public function __construct(
        public string $char
    ) {}

    public function jsonSerialize(): mixed
    {
        return [
            'ir' => 'Char',
            'char' => $this->char,
        ];
    }
}

readonly class Esc implements IRNode
{
    public function __construct(
        public string $type,
        public ?string $property = null
    ) {}

    public function jsonSerialize(): mixed
    {
        $data = [
            'ir' => 'Esc',
            'type' => $this->type,
        ];
        if ($this->property !== null) {
            $data['property'] = $this->property;
        }
        return $data;
    }
}

readonly class Group implements IRNode
{
    public function __construct(
        public IRNode $child,
        public bool $capturing,
        public ?string $name = null,
        public ?int $number = null
    ) {}

    public function jsonSerialize(): mixed
    {
        $data = [
            'ir' => 'Group',
            'child' => $this->child,
            'capturing' => $this->capturing,
        ];
        if ($this->name !== null) $data['name'] = $this->name;
        if ($this->number !== null) $data['number'] = $this->number;
        return $data;
    }
}

readonly class Alt implements IRNode
{
    /** @param IRNode[] $alternatives */
    public function __construct(
        public array $alternatives
    ) {}

    public function jsonSerialize(): mixed
    {
        return [
            'ir' => 'Alt',
            'alternatives' => $this->alternatives,
        ];
    }
}

readonly class Anchor implements IRNode
{
    public function __construct(
        public string $type
    ) {}

    public function jsonSerialize(): mixed
    {
        return [
            'ir' => 'Anchor',
            'type' => $this->type,
        ];
    }
}

readonly class Range implements IRNode
{
    public function __construct(
        public string $from,
        public string $to
    ) {}

    public function jsonSerialize(): mixed
    {
        return [
            'ir' => 'Range',
            'from' => $this->from,
            'to' => $this->to,
        ];
    }
}

readonly class BackRef implements IRNode
{
    public function __construct(
        public ?int $index = null,
        public ?string $name = null
    ) {}

    public function jsonSerialize(): mixed
    {
        $data = ['ir' => 'BackRef'];
        if ($this->index !== null) $data['index'] = $this->index;
        if ($this->name !== null) $data['name'] = $this->name;
        return $data;
    }
}

readonly class LookAround implements IRNode
{
    public function __construct(
        public string $type, // "Lookahead", "Lookbehind"
        public bool $negated,
        public IRNode $child
    ) {}

    public function jsonSerialize(): mixed
    {
        return [
            'ir' => 'LookAround',
            'type' => $this->type,
            'negated' => $this->negated,
            'child' => $this->child,
        ];
    }
}
